File: src/pages/kualitas-air.js
const { invoke } = window.__TAURI__.core;

export function initKualitasAir() {
    console.log("🚀 Module Kualitas Air dimuat.");
    initFlatpicker();
    
    // Listener Tombol Simpan
    const saveBtn = document.getElementById('saveBtn');
    if (saveBtn) saveBtn.addEventListener('click', saveData);

    // Listener Tombol Hitung IP
    const btnHitungIP = document.getElementById('btnHitungIP');
    if (btnHitungIP) btnHitungIP.addEventListener('click', calculateIP);

    // Listener Tombol Upload
    const btnUpload = document.getElementById('btnUploadPdf');
    if (btnUpload) {
        btnUpload.addEventListener('click', handlePdfUpload);
    }
}

/**
 * Fitur: Upload PDF via Rust Native Dialog
 */
async function handlePdfUpload() {
    try {
        console.log("Meminta Rust membuka dialog file...");
        document.body.style.cursor = 'wait';

        // 1. Panggil Rust untuk baca PDF
        const data = await invoke('import_pdf');

        document.body.style.cursor = 'default';

        if (data) {
            console.log("📄 Data Diterima dari Rust:", data);
            
            // 2. Isi Form Otomatis
            fillForm(data);

            alert("✅ Data berhasil diekstrak dan dimasukkan ke form!");
        }

    } catch (error) {
        document.body.style.cursor = 'default';
        
        if (error.includes("dibatalkan")) {
            console.log("User membatalkan pemilihan file.");
            return;
        }

        console.error("PDF Error:", error);
        alert("❌ Gagal: " + error);
    }
}

/**
 * Helper: Memetakan Data Rust ke Input HTML
 */
function fillForm(data) {
    // Fungsi kecil untuk set nilai jika element ada & datanya tidak null
    const setVal = (id, val) => {
        const el = document.getElementById(id);
        if (el && val !== null && val !== undefined) {
            el.value = val;
            
            // Visual feedback (opsional: kedip kuning sebentar)
            el.style.backgroundColor = "#fff3cd";
            setTimeout(() => { el.style.backgroundColor = ""; }, 1000);
        }
    };

    // --- MAPPING FIELD RUST -> HTML ID ---
    
    // A. Parameter Fisika
    setVal('temperatur', data.temperatur);
    setVal('konduktivitas', data.konduktivitas); // DHL
    setVal('kekeruhan', data.kekeruhan);
    setVal('tds', data.tds);
    setVal('tss', data.tss);
    setVal('warna', data.warna);

    // B. Parameter Kimia
    setVal('ph', data.ph);
    setVal('oksigen', data.oksigen); // DO
    setVal('bod', data.bod);
    setVal('cod', data.cod);
    setVal('amoniak', data.amoniak);
    setVal('nitrat', data.nitrat);
    setVal('nitrit', data.nitrit);
    setVal('fosfat', data.fosfat);
    setVal('deterjen', data.deterjen);
    setVal('minyakDanLemak', data.minyakDanLemak); // Perhatikan beda snake_case ke camelCase
    setVal('fenol', data.fenol);
    
    // C. Parameter Logam & Anorganik
    setVal('klorida', data.klorida);
    setVal('belerang', data.belerang);
    setVal('sianida', data.sianida);
    setVal('fluorida', data.fluorida);
    setVal('arsen', data.arsen);
    setVal('besi', data.besi);
    setVal('mangan', data.mangan);
    setVal('tembaga', data.tembaga);
    setVal('merkuri', data.merkuri);

    // D. Mikrobiologi
    setVal('totalColiform', data.totalColiform);

    // Catatan: Metadata (Nama Pos, dll) belum diekstrak oleh Rust, 
    // jadi belum kita map di sini.
}

// --- FUNGSI STANDAR LAINNYA ---

function getFormData() {
    const getVal = (id) => {
        const el = document.getElementById(id);
        if (!el) return null;
        const val = el.value.trim();
        if (val === "") return null;
        const stringFields = ['namaPos', 'das', 'wilayahSungai', 'provinsi', 'kabupaten', 'elevasiPos', 'pelaksana', 'kecamatan', 'laboratorium', 'sungai', 'desa', 'koordinatGeografis', 'sampleDate', 'sampleTime', 'statusIP', 'statusStoret'];
        if (stringFields.includes(id)) return val;
        const num = parseFloat(val.replace(',', '.'));
        return isNaN(num) ? null : num;
    };
    return {
        nama_pos: getVal('namaPos'), das: getVal('das'), wilayah_sungai: getVal('wilayahSungai'), provinsi: getVal('provinsi'), kabupaten: getVal('kabupaten'), tahun: getVal('tahun'), elevasi_pos: getVal('elevasiPos'), pelaksana: getVal('pelaksana'), kecamatan: getVal('kecamatan'), laboratorium: getVal('laboratorium'), sungai: getVal('sungai'), desa: getVal('desa'), koordinat_geografis: getVal('koordinatGeografis'),
        tanggal_sampling: getVal('sampleDate'), waktu_sampling: getVal('sampleTime'),
        temperatur: getVal('temperatur'), konduktivitas: getVal('konduktivitas'), kekeruhan: getVal('kekeruhan'), oksigen: getVal('oksigen'), ph: getVal('ph'), tds: getVal('tds'), tss: getVal('tss'), warna: getVal('warna'), klorida: getVal('klorida'), amoniak: getVal('amoniak'), nitrat: getVal('nitrat'), nitrit: getVal('nitrit'), fosfat: getVal('fosfat'), deterjen: getVal('deterjen'), arsen: getVal('arsen'), besi: getVal('besi'), mangan: getVal('mangan'), tembaga: getVal('tembaga'), merkuri: getVal('merkuri'), sianida: getVal('sianida'), fluorida: getVal('fluorida'), belerang: getVal('belerang'), cod: getVal('cod'), bod: getVal('bod'), fenol: getVal('fenol'), debit: getVal('debit'), minyakDanLemak: getVal('minyakDanLemak'), totalColiform: getVal('totalColiform'),
        nilaiIp: getVal('metodeIndeksPencemaran'), statusIp: getVal('statusIP'), nilaiStoret: getVal('metodeStoret'), statusStoret: getVal('statusStoret')
    };
}

async function calculateIP() {
    const payload = getFormData();
    try {
        const [score, status] = await invoke('calculate_ip_preview', { data: payload });
        const inputNilai = document.getElementById('metodeIndeksPencemaran');
        const inputStatus = document.getElementById('statusIP');
        if (inputNilai) inputNilai.value = score;
        if (inputStatus) inputStatus.value = status;
        if (status.includes("Baik") || status.includes("Memenuhi")) inputStatus.style.color = "green";
        else if (status.includes("Ringan")) inputStatus.style.color = "#f39c12"; 
        else inputStatus.style.color = "red";
    } catch (error) { console.error(error); alert("❌ Gagal: " + error); }
}

async function saveData() {
    const payload = getFormData();
    try {
        const response = await invoke('submit_kualitas_air', { data: payload });
        alert("✅ SUKSES: " + response);
    } catch (error) { console.error(error); alert("❌ ERROR: " + error); }
}

function initFlatpicker() {
    if (typeof flatpickr !== 'undefined') {
        flatpickr("#sampleDate", { dateFormat: "d-m-Y" });
        flatpickr("#sampleTime", { enableTime: true, noCalendar: true, dateFormat: "H:i", time_24hr: true });
    }
}

File: src-tauri/src/models/kualitas_air.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use chrono::NaiveDateTime;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct KualitasAirRecord {
    pub id: Option<i64>,
    
    // --- METADATA HEADER ---
    pub nama_pos: Option<String>,
    pub das: Option<String>,
    pub wilayah_sungai: Option<String>,
    pub provinsi: Option<String>,
    pub kabupaten: Option<String>,
    pub tahun: Option<i32>,
    pub elevasi_pos: Option<String>,
    pub pelaksana: Option<String>,
    pub kecamatan: Option<String>,
    pub laboratorium: Option<String>,
    pub sungai: Option<String>,
    pub desa: Option<String>,
    pub koordinat_geografis: Option<String>,

    // --- WAKTU SAMPLING ---
    pub tanggal_sampling: Option<String>,
    pub waktu_sampling: Option<String>,
    
    // --- PARAMETER KUALITAS AIR ---
    // 1. Variabel Umum
    pub temperatur: Option<f64>,
    pub konduktivitas: Option<f64>,
    pub kekeruhan: Option<f64>,
    pub oksigen: Option<f64>, // DO
    pub ph: Option<f64>,
    pub tds: Option<f64>,
    pub tss: Option<f64>,
    pub warna: Option<f64>,

    // 2. Major Ion
    pub klorida: Option<f64>,

    // 3. Nutrient
    pub amoniak: Option<f64>,
    pub nitrat: Option<f64>,
    pub nitrit: Option<f64>,
    pub fosfat: Option<f64>,
    pub deterjen: Option<f64>,

    // 4. Logam Berat
    pub arsen: Option<f64>,
    pub besi: Option<f64>,
    pub mangan: Option<f64>,
    pub tembaga: Option<f64>,
    pub merkuri: Option<f64>,

    // 5. Anorganik Lain
    pub sianida: Option<f64>,
    pub fluorida: Option<f64>,
    pub belerang: Option<f64>,

    // 6. Organik
    pub cod: Option<f64>,
    pub bod: Option<f64>,
    
    #[serde(rename = "minyakDanLemak")]
    #[sqlx(rename = "minyak_dan_lemak")] 
    pub minyak_dan_lemak: Option<f64>,

    pub fenol: Option<f64>,

    // 7. Mikrobiologi
    #[serde(rename = "totalColiform")]
    #[sqlx(rename = "total_coliform")]
    pub total_coliform: Option<f64>,

    // 8. Lain-lain
    pub debit: Option<f64>,

    // --- HASIL KALKULASI (OUTPUT) ---
    // Kita simpan agar bisa di-query & report tanpa hitung ulang
    
    #[sqlx(rename = "nilai_ip")]
    #[serde(rename = "nilaiIp")]
    pub nilai_ip: Option<f64>, // Contoh: 5.2

    #[sqlx(rename = "status_ip")]
    #[serde(rename = "statusIp")]
    pub status_ip: Option<String>, // Contoh: "Cemar Sedang"

    #[sqlx(rename = "nilai_storet")]
    #[serde(rename = "nilaiStoret")]
    pub nilai_storet: Option<f64>, // Contoh: -18.0

    #[sqlx(rename = "status_storet")]
    #[serde(rename = "statusStoret")]
    pub status_storet: Option<String>, // Contoh: "Kelas C"

    // --- SYSTEM METADATA ---
    pub created_at: Option<NaiveDateTime>,
}